### ISSUE TITLE
Feature: libmypaint for brush tool in raster levels

### ISSUE THREAD

**Issue by blurymind** [opentoonz/opentoonz#417]
Libmypaint is a popular brush library used by mypaint and now also available for gimp. Pencil2d devs did some work to make it easier to integrate in qt apps. I thought that it might be useful to have a link to it here:

https://github.com/feeef/QTMyPaint

This is a big task, but it will make the brush tool extremely powerful.
We can take the best from the open source world :)

Btw i will post some demo videos of mypaint in action later.

Here is libmypaint's original github for the latest version:
https://github.com/mypaint/libmypaint

**Comment by blurymind** [comment]
Libmypaint is a very powerful open source brush engine. It had a huge set of brush presets created for it. Its now a part of gimp and mypaint, but it is also ported to blender, krita and pencil2d among others. Here are some demo videos:

https://youtu.be/MLMqOyxUPvg

https://youtu.be/T-_gofxpgNQ

http://www.davidrevoy.com/article142/ressource-mypaint-brushes

http://www.davidrevoy.com/article55/mypaint-v4-brushkit

https://youtu.be/hzUwPF1_6H8

**Comment by blurymind** [comment]
Here is something relevant you should consider:
https://plus.google.com/+gimp/posts/agohJvtj4Sm

During +Libre Graphics Meeting in Madrid we met awesome folks from +mypaint, Krita, and Tupi teams and discussed how we could simplify sharing resources like brushes. So we came up with a proposal for a simple ZIP-based file format for distribution. +Boudewijn Rempt of Krita fame posted the first version of the proposal for public discussion last night.

Later on we could also use this file format for distribution of other types of 3rd party resources (gradients, palettes etc.).

We are also considering adding support for MyPaint brushes in either 2.10 or 3.0, depending on the workload. This is now easy, since MyPaint team recently abstracted their brush engine into a nice simple library.﻿
https://lists.freedesktop.org/archives/create/2013-April/004616.html

They already added libmypaint to gimp btw.

**Comment by turtletooth** [comment]
It is GPL licensed.  I love the idea, but the licenses are incompatible.

**Comment by Banbury** [comment]
QTMyPaint is GPL. libmypaint isn't. 

So writing your own wrapper is still an option. I've had a short look at QTMyPaint and it doesn't look like brain surgery to me.

**Comment by turtletooth** [comment]
Awesome!  Good to know!

**Comment by blurymind** [comment]
qtmypaint is just a qt implementation for the interface, right?
So it is good to look at the code there as an example or proof of concept. In order to fit the mypaint brush lib within opentoonz brush tool's design (raster layer mode), you are going to need to do a much different interface.

**Comment by blurymind** [comment]
Here I created a mocup of how libmypaint brushes can work in the style editor's gui. All libmypaint brushes could be in a 'myPaint brushes' tab like this:
![mypaintbrushes](https://cloud.githubusercontent.com/assets/6495061/15949038/68d089d0-2e9c-11e6-829b-4d79b12f5733.png)
This mockup is a followup to #429 as it introduces brush categories or brush sets drop down menu  to the style editor. something that needs to happen before adding libmypaint or other brush engines to OT imo.

Another needed component is ofcourse the creation and authoring of mypaint brushes. This is where the 'Settings' tab can serve it's job perfectly:
![mypaintbrushes-settings](https://cloud.githubusercontent.com/assets/6495061/15949197/818e721a-2e9d-11e6-83a8-b58723b52c11.png)

It is good to note that libmypaint has a brush stabilizer built into it. This is something that opentoonz does not have yet. @walkerka has implemented something for the vector brush tool that adds stabilization. So if in the future libmypaint finds its way in opentoonz, there would be no need for @walkerka to make his vector brush tool stabilization functionality also work on the raster brush tool.

**Comment by feeef** [comment]
Hi, my name is François, I am the original author of the QTMypaint lib.

I choose the GPLv3 license as a wish to support the copyleft philosophy but I realize that many free software projects like yours cannot benefit from my code, which is far from my intentions.

I plan to change its license to a less restrictive one so you may reuse the code as you wish.

I have started a discussion on this issue here : https://github.com/feeef/qtmypaint/issues/8

Thank you.

**Comment by feeef** [comment]
I have now changed the license of QTMypaint to "Modified BSD" so feel free to use it as you wish within OpenToonz.

Thank you! :)

**Comment by blurymind** [comment]
@feeef thank you! :)
And also thank you for integrating libmypaint into pencil2d.
Would you be interested in also integrating it into opentoonz?

@walkerka @turtletooth @shun-iwasawa 
Perhaps now that there is no licensing issues, someone would be interested to do it for the standard raster level mode of the brush tool?

**Comment by morevnaproject** [comment]
@feeef Thank you! ^__^

**Comment by turtletooth** [comment]
@feeef I haven't had time to look very closely at this, but it seems it would work great on the raster layers in toonz.  How hard has this been to implement in Pencil2D?

**Comment by adamearle** [comment]
Personal opinion here, but it would be great to intergrate 'pencil 2D'
timeline into open toonz.

I think that this may save on space in the interface as you can fit more
keys horizontally than vertically.

People at least then have an option on how they want to work.

'Mypaint' integration is truely exciting isn't it.

Adam

**Comment by blurymind** [comment]
@adamearle that is for another feature request. Horizontal timeline has already been requested in the past
https://github.com/opentoonz/opentoonz/issues/503

**Comment by feeef** [comment]
@turtletooth, implementing the brush engine is quite strainght forward with qtmypaint. It uses a  QGraphicsScene object with tiles of QGraphicsItems. Even if the OpenToonz canvas is not a QGraphicsScene, you can set up your own virtual QGraphicsScene and display its content to the OpenToonz canvas. qtmypaint has methods to get and set your drawing anytime as a QImage.

The only problem I have encontered on Pencil2D is that it has an infinite canvas and I still haven't managed to find a way to safely implement zooming, rotating and translating. QTMyPaint uses a predefined canvas size in order to calculate the number of required tiles but I made it in a way that you can easily change this size and re-position your image at any time.

@blurymind, unfortunately, I am currently too busy to implement this feature in OpenToonz and if I get the time, I'd rather finish working on the Pencil2D implementation. Pencil2D is not as sophisticated as OpenToonz but I like it a lot and wish to have it work correctly.

I am currently working at Purism and may get a Librem 11 when it will be released at the end of the year. So I will try to get the time to work on making Pencil2D's UX optimized for this kind tablet computer.

However, I have managed to build OpenToonz on my Debian box, it seems to be amazingly powerful and I will follow its updates with a lot of interest! :)

Cheers,
Francois.

**Comment by blurymind** [comment]
@feeef  That is quite interesting. Thank you for clarifying things. It's unfortunate you dont have the time, but you could still continue assisting the implementation with good advice whenever  :) 
Mypaint also has an unlimited canvas and libmypaint comes from it, so how is the unlimited canvas in pencil2d different from the one in mypaint?

I am quite happy to see it already partially working in pencil2d and even more eager to see in in opentoonz - because OT is a much more mature piece of software with real production houses such as Ghibli having used it already. 

That said I still like to use pencil2d and was even involved in creating it's facebook and google+ community groups. It's a wonderful tool for flipbook animation and seeing that it gets much deserved developer attention makes me infinitely happy.
As an owner of a samsung galaxy note, I would be quite excited to see that running on android.

**Comment by feeef** [comment]
@blurymind, MyPaint uses an infinite canvas but the visible area is always limited by the size of the window. libmypaint only handles this limited area, so zooming, translating and rotating has to be implemented in the application.

**Comment by feeef** [comment]
Well, what I just said is not exactly true. libmypaint can handle more than the visible area but will always require a size for the "drawable" area. In the case of an infinite canvas, we usually limit this drawable area to the visible area.

**Comment by morevnaproject** [comment]
I have posted a bounty for this feature - https://www.bountysource.com/issues/34932107-feature-libmypaint-for-brush-tool-in-raster-levels

I would appreciate if you help to spread a word about it. Maybe someone will want to join me and pitch the reward amount. In this case there are higher chances to get some developer interested to implement this. ^__^

![43e2790e-8d78-11e6-9500-d84caf2a9fa1](https://cloud.githubusercontent.com/assets/332868/19807797/f8dfe8f4-9d4b-11e6-811c-11c6ed095b5c.png)

**Comment by blurymind** [comment]
Thank you :) !!

**Comment by feeef** [comment]
Thanks a lot @morevnaproject !

**Comment by morevnaproject** [comment]
The bounty is $10 now! + we have a developer who is willing to implement this feature. I have made a blog post - https://morevnaproject.org/2017/01/14/mypaint-brushes-opentoonz-goes-bounty/
Come on, people, let's make this feature into OpenToonz! For Linux folks we will deliver binary builds, so they will be able to test feature earlier than anyone else. ^__^

**Comment by blurymind** [comment]
He will do it only for 10 dollars? :o
This is so getting funded in no time!

**Comment by morevnaproject** [comment]
> He will do it only for 10 dollars? :o
The bounty goal is set to $150, please check the bounty page.

![screen-10](https://cloud.githubusercontent.com/assets/332868/21980028/15b5bf7e-dc14-11e6-9621-f583c0e3eaa7.png)

**Comment by blurymind** [comment]
I think its reasonable!

Its worth posting this to:
The opentoonz google forum
The google plus opentoonz group
Twitter
 blender forum's thread
Etc

**Comment by morevnaproject** [comment]
I have posted everywhere I could. Any help is appreciated here. ^__^

Already posted to:
- [ ] The opentoonz google forum 
- [ ] The google plus opentoonz group 
- [ ] Twitter

**Comment by HulluMel** [comment]
This would be an amazing addition to OT. I make tutorials for OpenToonz and I would love to promote the funding. Would it be ok if I used the mockups to illustrate my points?

**Comment by blurymind** [comment]
Yes it will be fine :)
Go ahead

**Comment by morevnaproject** [comment]
@HulluMel That would be awesome! Please let me know the link to your video - I will be happy to share it with our followers.

Thanks to Alexey Ignatyev, @turtletooth, Animationista and Stepan Krani  - the bounty is at 37% now!
![screen-14](https://cloud.githubusercontent.com/assets/332868/22336367/968905c2-e414-11e6-85c6-97e17cbe5d71.png)

https://www.bountysource.com/issues/34932107-feature-libmypaint-for-brush-tool-in-raster-levels

**Comment by morevnaproject** [comment]
Much thanks to @HulluMel  - we are at 53% now!
![screen-12](https://cloud.githubusercontent.com/assets/332868/22410772/0569aaca-e6cf-11e6-8e69-dd48514df163.png)
Bounty page: https://www.bountysource.com/issues/34932107-feature-libmypaint-for-brush-tool-in-raster-levels

**Comment by morevnaproject** [comment]
63% towards the funding goal. Much thanks to Dorne for support!
Bounty page: https://www.bountysource.com/issues/34932107-feature-libmypaint-for-brush-tool-in-raster-levels

**Comment by morevnaproject** [comment]
I have made a mention of the bounty in this article - https://opensource.com/article/17/2/opentoonz-2d-animation-software

**Comment by turtletooth** [comment]
The bounty just reached 100%!! Thank you so much @MasterTurmeric

**Comment by morevnaproject** [comment]
Much thanks to @MasterTurmeric for the generous contribution!  The goal is reached and our developer @blackwarthog starts to work on it today!
![screen-13](https://cloud.githubusercontent.com/assets/332868/22767956/6b19ac1e-eeb1-11e6-881d-4f381b0416b7.png)
Much thanks to everyone for participation. ^__^

**Comment by turtletooth** [comment]
I'm really excited for this.  @blackwarthog - if you need any testing, I'll be glad to help.

**Comment by blurymind** [comment]
@morevnaproject Thank you and congratulations on the funding!
It's exciting to get konstantin on board with this project, as he had done some amazing work on synfig

**Comment by eVoLuTiVeMEDIA** [comment]
QTMypaint lib is working great.
Bon travail, François @feeef 
(I'm going to test pencil with mypaint lib right now, thanks)

**Comment by feeef** [comment]
thank you @eVoLuTiVeMEDIA!
Pencil2D with mypaint is still just a preview and far from being complete. It should work when drawing but I had trouble making zoom and pan features. Unfortunately, I have been very busy. I spend my free time on developing a free licensed film production (http://ethiccinema.org) and didn't get the time to really work on finishing this integration. I wish Ethic Cinema to sponsor this kind of development in the future.

**Comment by blackwarthog** [comment]
Good news!

More than week i spent in coding. And now i can show some results. You may see it at attached screenshot and in [source code](https://github.com/blackwarthog/opentoonz/tree/brush) in my repository.

![mypaint-brush-screenshot](https://cloud.githubusercontent.com/assets/5143743/23125978/8bf231de-f7af-11e6-8fb0-d0d4007ab1ab.png)

So, the first, that's already done:
 - MyPaint brushes now works in OpenToonz.

And that's remains:
 - increase speed - rendering of the dab implemented without any optimizations, yet;
 - improve interpolation, because mypaint doesn't use spline interpolation of stylus motion, but all other OpenToonz brushes does it;
 - configure of building process for all supported platforms;

Much thanks! I'll go (or be?) back to coding...

**Comment by turtletooth** [comment]
@blackwarthog Thanks so much.  Can't wait to try it out.

**Comment by adamearle** [comment]
WOW that was fast can't wait to play with this once its done. What about 
some brush engine features from Krita next.

Ad

**Comment by morevnaproject** [comment]
Dear friends! @blackwarthog made first testing of OpenToonz with MyPaint brushes feature! Available for Windows and Linux platforms - https://morevnaproject.org/2017/03/16/opentoonz-mypaint-brushes-available-download/

Note: Besides of providing our usual Linux builds, @blackwarthog put a special efforts into preparing Windows builds for our users. Preparation of those Windows builds are related with some configuration changes required for OpenToonz build system - see PR #1092. Unfortunately, it seems there is a lack of interest in merging this PR. If this PR will remain unmerged, then for us it would be much harder to provide you Windows builds in the future (because we will need to do additional efforts every time when OpenToonz build scripts changes). If you are interested in future Windows builds from us, then I would appreciate your help in convincing developers to merge this PR. Thank you! ^__^

**Comment by turtletooth** [comment]
@morevnaproject I can't really make any decisions about the build system or the PR, but I would gladly create portable windows versions for testing purposes.

**Comment by morevnaproject** [comment]
@turtletooth Thank you for the proposal! But in order to fix some OS-related issues we need to have possibility to build it ourselves. We need to do debugging. ^__^

**Comment by artisteacher** [comment]
I hope this will be available for OS X as well!

**Comment by turtletooth** [comment]
First of all- thank you so much for this! It is so fun to be able to draw with the MyPaint brushes in OpenToonz.

Here's some quick feedback on functionality/usability.

Are there any plans to bring the brush settings to the UI? Not having access to settings limits customization.

I almost feel like the MyPaint tab and MyPaint brush settings could be its own panel that could be docked. Right now, there is a lot of clicking back and forth between tabs to adjust color and switch brushes.

The brushes sometimes seem too big. A brush is never able to be one pixel. This may be a more general OpenToonz problem.

One probably crazy thought that crossed my mind is removing the standard raster brush entirely and using only MyPaint brushes. This would eliminate some confusion for users about what type of raster brush they are using. Another similar thought is pulling the MyPaint brush to its own tool.

Sometimes the brush circle preview doesn't accurately represent how big the brush will be. Example: use the "Brush" brush at the top, set the brush size to 100. The cursor is big but not huge, but if you draw- wow - it's huge. This also works backwards, a brush set at one, never only draws one pixel.

Does hardness affect any MyPaint brushes? It doesn't seem to. That is part of why I mentioned splitting the two brushes- to remove confusion. Opacity doesn't seem to affect all brushes also.

Again thank you for all the great work. I am so excited to have this in OpenToonz.

**Comment by tGuMedia** [comment]
@turtletooth 
> One probably crazy thought that crossed my mind is removing the standard raster brush entirely and using only MyPaint brushes. This would eliminate some confusion for users about what type of raster brush they are using. Another similar thought is pulling the MyPaint brush to its own tool.

I don't think pulling the standard brush by its entirety is at all a good idea because personally,  i wouldn't really benefit from having the MyPaint brushes, even though it's really cool that It's starting to work, i don't want to be forced into using them.

Separating the standard brushtool and the mypaint brushtool seems to be a fine idea though.

**Comment by turtletooth** [comment]
What about moving the MyPaint brush to the other brush tool.  Currently that only works for Toonz Raster levels and does nothing for the other level types.

**Comment by tGuMedia** [comment]
@turtletooth yeah, sure the Paint Brush serves no real purpose anyway.

also i haven't been able to test yet but i'm gonna assume that the goal for this is for it to only on the normal raster levels, am i correct?

**Comment by konero** [comment]
Actually the Paint Brush Tool is used extensively in paint tasks as it allows you to paint over **only lines** etc - we use it all the time.

**Comment by turtletooth** [comment]
@konero - the Paint Brush tool doesn't do anything on raster levels.  But it seems this is a non-issue.  From a reply I received from @blackwarthog on his repository - it seems that on raster levels, the standard brush always uses MyPaint brushes - using MyPaint to mimic the old brush when a "non-MyPaint brush" is being used.

@tGuMedia Yes- MyPaint will only affect raster levels.

**Comment by RodneyBaker** [comment]
I would caution against thinking of MyPaint as a replacement for the
standard Paintbrush Tool.  There are brushes for use with Raster (Toonz
Raster that is) that convert quite well to vector but few (if any) of the
My Paint brushes do.

Its easy to confuse the various types of levels.
MyPaint only works on generic Raster levels.  I don't think there are plans
to extend it to Toonz Raster levels (in other words I don't know)

The underlying issue as far as I can tell is that the standard 'Raster'
brushes that come with OpenToonz convert very nicely to Vector (and fill
too!) but lines created with MyPaint do not.

It may be that as of yet I've just not discovered what the equivalent brush
in MyPaint is that will produce the same results as the standard brush set.

End word of caution.  :)

I confess that until recently I've rarely used generic Raster level
preferring to use Toonz Vector and Toonz Raster levels (in that order).
I was leaning toward use of Toonz Raster in order to get better fillable
shapes for use with conversions to Toonz vector.

Having said all that I really love the MyPaint addition and plan to use the
heck out of it!

**Comment by blackwarthog** [comment]
Allow me to give some technical notes about MyPaint-brush implementation.

As @turtletooth ( [and me :)](https://github.com/blackwarthog/opentoonz/issues/1#issuecomment-287421955) ) said analogue of standard brush still presents in OpenToonz, just select any non-mypaint style to enable it.

Technically original OpenToonz brush for RasterLevel is an separate tool named FullColorBrush. Seems it's the rule of OpenToonz - to minimize tools count in GUI. FullColorBrush is associates with button "BrushTool" and "RasterLayer". Other tools associates this button with other layers. I've modified FullColorBrush tool to implement MyPaint brush.

It's possible to implement MyPaint brush for ToonzRasterLayer. But ToonzRasterLayer impose some restrictions to brush effects - so color dynamics, gradients, color callbacks and brushes with special blend methods will not working as expected. Anyway exists a lot of MyPaint brushes, which will work fine at ToonzRasterLayer.

And MyPaint-brush at ToonzVectorLayer is not so impossible as expected :)

**Comment by blurymind** [comment]
Imo this should be reviewed and merged asap
Mypaint's brushlib is much nicer to use for sketching and adds the ability to do proper painting in opentoonz - without nullifying or breaking any of the currently existing features or tools.

One big missing feature of course is the ability to author mypaint brushes in opentoonz - that could perhaps be added gradually?

**Comment by morevnaproject** [comment]
Important to everyone who downloaded Windows build (with MyPaint brushes)! There is a nasty bug, which corrupts contents of raster levels. @blackwarthog have fixed this bug now and updated builds are available for download. Please see details here - https://morevnaproject.org/2017/03/23/important-fix-opentoonz-windows/

**Comment by morevnaproject** [comment]
Hello everyone! @blackwarthog made a bunch of fixes and improvement to MyPaint functionality in OpenToonz. Please read  detailed description and get updated builds - https://morevnaproject.org/2017/04/09/open-toonz-builds-updated/

**Comment by blurymind** [comment]
hope to see this merged soon. Thank you very much for this great contribution!

**Comment by turtletooth** [comment]
Any news on when we might be able to build this with Windows locally?

**Comment by morevnaproject** [comment]
@turtletooth 
By this moment @blackwarthog added support for OSX build. Now he is started with final stage  of adding support for Windows MSVC. 

Also, we eagerly waiting for this pull request to be merged - https://github.com/opentoonz/opentoonz/pull/1118. Without it our Windows build of OT will not save files correctly. ^__^

**Comment by turtletooth** [comment]
#1118 Merged.  Thanks @morevnaproject and @blackwarthog for your great work.

**Comment by morevnaproject** [comment]
Awesome! Thank you! ^__^

**Comment by blurymind** [comment]
@blackwarthog @morevnaproject @shun-iwasawa @turtletooth 
Thank you all for making this possible! Opentoonz will be so much nicer to use for drawing now! :D 
libmypaint is an incredibly fun to use brush engine!

**Comment by morevnaproject** [comment]
Much thanks to everyone who supported the bounty and made this possible. ^__^
